File: src-tauri/src/store.rs
use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};

use crate::site::Fact;

#[derive(Serialize, Deserialize, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct Settings {
    #[serde(default)]
    pub username: String,
    #[serde(default)]
    pub password: String,
    #[serde(default)]
    pub video_dir: Option<String>,
    #[serde(default)]
    pub player: Option<String>,
    /// Смотреть прямо с CDN, не сохраняя файл на диск.
    #[serde(default)]
    pub stream_by_default: bool,
}

#[derive(Serialize, Deserialize, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct Entry {
    pub url: String,
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub original: String,
    #[serde(default)]
    pub poster: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub genres: Vec<String>,
    #[serde(default)]
    pub facts: Vec<Fact>,
    #[serde(default)]
    pub watched: Vec<String>,
    #[serde(default)]
    pub last_player: Option<String>,
    #[serde(default)]
    pub last_quality: Option<String>,
    #[serde(default)]
    pub added_at: u64,
}

#[derive(Serialize, Deserialize, Default)]
struct OldLink {
    url: String,
    #[serde(default)]
    title: String,
}

fn home_dir() -> Option<PathBuf> {
    std::env::var_os(if cfg!(windows) { "USERPROFILE" } else { "HOME" }).map(PathBuf::from)
}

pub fn config_dir() -> Result<PathBuf> {
    let base = if cfg!(windows) {
        std::env::var_os("APPDATA")
            .map(PathBuf::from)
            .or_else(|| home_dir().map(|h| h.join("AppData").join("Roaming")))
    } else {
        std::env::var_os("XDG_CONFIG_HOME")
            .map(PathBuf::from)
            .or_else(|| home_dir().map(|h| h.join(".config")))
    }
    .ok_or_else(|| anyhow!("не найдена домашняя папка пользователя"))?;
    let dir = base.join("animejoya");
    std::fs::create_dir_all(&dir)?;
    Ok(dir)
}

fn read_json<T: for<'a> Deserialize<'a> + Default>(path: &Path) -> T {
    std::fs::read_to_string(path)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or_default()
}

fn write_json<T: Serialize>(path: &Path, value: &T) -> Result<()> {
    std::fs::write(path, serde_json::to_string_pretty(value)?)?;
    Ok(())
}

#[cfg(unix)]
fn restrict(path: &Path) {
    use std::os::unix::fs::PermissionsExt;
    let _ = std::fs::set_permissions(path, std::fs::Permissions::from_mode(0o600));
}
#[cfg(not(unix))]
fn restrict(_: &Path) {}

pub fn now() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

pub fn load_settings() -> Result<Settings> {
    Ok(read_json(&config_dir()?.join("config.json")))
}

pub fn save_settings(s: &Settings) -> Result<()> {
    let path = config_dir()?.join("config.json");
    write_json(&path, s)?;
    restrict(&path);
    Ok(())
}

fn library_path() -> Result<PathBuf> {
    Ok(config_dir()?.join("library.json"))
}

/// Первый запуск после CLI-версии: подхватываем старый links.json.
pub fn load_library() -> Result<Vec<Entry>> {
    let path = library_path()?;
    if path.exists() {
        return Ok(read_json(&path));
    }
    let old: Vec<OldLink> = read_json(&config_dir()?.join("links.json"));
    Ok(old
        .into_iter()
        .map(|l| Entry {
            url: l.url,
            title: l.title,
            added_at: now(),
            ..Default::default()
        })
        .collect())
}

pub fn save_library(items: &[Entry]) -> Result<()> {
    write_json(&library_path()?, &items)
}

pub fn video_dir(s: &Settings) -> PathBuf {
    if let Some(d) = std::env::var_os("ANIMEJOYA_DIR") {
        return PathBuf::from(d);
    }
    if let Some(d) = s.video_dir.as_ref().filter(|d| !d.trim().is_empty()) {
        return PathBuf::from(d);
    }
    home_dir()
        .unwrap_or_default()
        .join("Videos")
        .join("AnimeJoy")
}

/// `.../5499-o-moem-pererozhdenii-v-sliz-4-sezon.html` -> `5499-o-moem-...`
pub fn slug(url: &str) -> String {
    url.rsplit('/')
        .next()
        .unwrap_or(url)
        .trim_end_matches(".html")
        .to_string()
}

/// «21 серия» -> «21», иначе безопасное имя из заголовка.
pub fn ep_tag(title: &str) -> String {
    let num: String = title.chars().take_while(|c| c.is_ascii_digit()).collect();
    if num.is_empty() {
        title
            .chars()
            .map(|c| if c.is_alphanumeric() { c } else { '_' })
            .collect()
    } else {
        format!("{num:0>2}")
    }
}

pub fn normalize_url(url: &str) -> String {
    let u = url.trim();
    if u.starts_with("http") {
        u.to_string()
    } else {
        format!("https://{}", u.trim_start_matches("//"))
    }
}

File: server/store.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import type { z } from "zod";
import { Entry, Settings } from "./schema.ts";

const DEFAULTS: Settings = {
  username: "",
  password: "",
  preloadNext: true,
};

export function configDir(): string {
  const home = os.homedir();
  const base =
    process.platform === "win32"
      ? process.env.APPDATA ?? path.join(home, "AppData", "Roaming")
      : process.env.XDG_CONFIG_HOME ?? path.join(home, ".config");
  const dir = path.join(base, "animejoya");
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

function readJson(file: string): unknown {
  try {
    return JSON.parse(fs.readFileSync(file, "utf8"));
  } catch {
    return undefined;
  }
}

/// Через временный файл: оборванная запись не должна съесть библиотеку.
function writeJson(file: string, value: unknown, mode?: number): void {
  const tmp = `${file}.tmp`;
  fs.writeFileSync(tmp, JSON.stringify(value, null, 2), { mode });
  fs.renameSync(tmp, file);
}

/// Поля проверяем по одному: битое значение откатывается к умолчанию, а лишние ключи
/// вроде `video_dir` из старых конфигов просто отбрасываются.
function lenient<S extends z.ZodObject>(schema: S, raw: unknown, base: z.infer<S>): z.infer<S> {
  const src = (raw !== null && typeof raw === "object" ? raw : {}) as Record<string, unknown>;
  const out: Record<string, unknown> = { ...base };
  for (const [key, field] of Object.entries(schema.shape)) {
    const r = field.safeParse(src[key]);
    if (r.success) out[key] = r.data;
  }
  return out as z.infer<S>;
}

export const now = (): number => Math.floor(Date.now() / 1000);

export function blank(url: string, title = ""): Entry {
  return {
    url,
    title,
    original: "",
    poster: "",
    description: "",
    genres: [],
    facts: [],
    watched: [],
    lastPlayer: null,
    lastQuality: null,
    total: 0,
    addedAt: now(),
  };
}

const settingsPath = (): string => path.join(configDir(), "config.json");
const libraryPath = (): string => path.join(configDir(), "library.json");

export function loadSettings(): Settings {
  return lenient(Settings, readJson(settingsPath()), DEFAULTS);
}

export function saveSettings(s: Settings): void {
  writeJson(settingsPath(), s, 0o600);
}

const toEntry = (raw: unknown): Entry | null => {
  const url = (raw as { url?: unknown } | null)?.url;
  return typeof url === "string" ? lenient(Entry, raw, blank(url)) : null;
};

/// Первый запуск после CLI-версии: подхватываем старый links.json.
export function loadLibrary(): Entry[] {
  const file = libraryPath();
  if (fs.existsSync(file)) {
    const raw = readJson(file);
    return (Array.isArray(raw) ? raw : []).map(toEntry).filter((e) => e !== null);
  }
  const old = readJson(path.join(configDir(), "links.json"));
  return (Array.isArray(old) ? old : []).map(toEntry).filter((e) => e !== null);
}

export function saveLibrary(items: Entry[]): void {
  writeJson(libraryPath(), items);
}

/// `.../5499-o-moem-pererozhdenii-v-sliz-4-sezon.html` -> `5499-o-moem-...`
export function slug(url: string): string {
  return (url.split("/").pop() ?? url).replace(/(\.html)+$/, "");
}

/// «21 серия» -> «21», иначе безопасное имя из заголовка.
export function epTag(title: string): string {
  const num = /^[0-9]+/.exec(title);
  if (num) return num[0].padStart(2, "0");
  return Array.from(title)
    .map((c) => (/[\p{L}\p{N}]/u.test(c) ? c : "_"))
    .join("");
}

export function normalizeUrl(url: string): string {
  const u = url.trim();
  return u.startsWith("http") ? u : `https://${u.replace(/^(\/\/)+/, "")}`;
}
